### PR TITLE
Fix `non-constant value` ICE (#90878)

### LINKED ISSUE #90878: thread 'rustc' panicked at 'attempt to subtract with overflow', rust/compiler/rustc_resolve/src/diagnostics.rs:458:49
### Code

```rust
 fn main() {
    |x: usize| [0; x];
    // (note the space before "fn")
}
```

When rustc is compiled with debug assertions, this results in a subtract with overflow on this line:

https://github.com/rust-lang/rust/blob/1b12d01903293453dd94aa170c82caf94415629f/compiler/rustc_resolve/src/diagnostics.rs#L458

And if compiled without debug assertions, you get a warning "Invalid span".

Introduced in https://github.com/rust-lang/rust/pull/80801

Affected versions: stable 1.56.1, nightly 2021-11-12

<details><summary><strong>Backtrace</strong></summary>
<p>

With debug assertions:
```
thread 'rustc' panicked at 'attempt to subtract with overflow', rust/compiler/rustc_resolve/src/diagnostics.rs:458:49
stack backtrace:
   0: rust_begin_unwind
             at /rustc/46b8e7488eae116722196e8390c1bd2ea2e396cf/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/46b8e7488eae116722196e8390c1bd2ea2e396cf/library/core/src/panicking.rs:106:14
   2: core::panicking::panic
             at /rustc/46b8e7488eae116722196e8390c1bd2ea2e396cf/library/core/src/panicking.rs:47:5
   3: rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::into_struct_error
   4: rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::report_error
   5: rustc_resolve::Resolver::resolve_ident_in_lexical_scope
   6: rustc_resolve::Resolver::resolve_path_with_ribs::{{closure}}
   7: rustc_resolve::Resolver::resolve_path_with_ribs
   8: rustc_resolve::late::LateResolutionVisitor::resolve_qpath_anywhere
   9: rustc_resolve::late::LateResolutionVisitor::smart_resolve_path_fragment
  10: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  11: rustc_resolve::late::LateResolutionVisitor::resolve_anon_const
  12: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  13: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_fn
  14: rustc_ast::visit::walk_expr
  15: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  16: rustc_resolve::late::LateResolutionVisitor::resolve_block
  17: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_fn
  18: rustc_ast::visit::walk_item
  19: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_item
  20: rustc_resolve::Resolver::resolve_crate::{{closure}}
  21: rustc_resolve::Resolver::resolve_crate
  22: rustc_interface::passes::configure_and_expand
  23: rustc_interface::queries::Queries::expansion
  24: rustc_interface::interface::run_compiler::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Without debug assertions, it only shows a warning "Invalid span":

```
   Compiling playground v0.0.1 (/playground)
WARN rustc_errors::emitter Invalid span src/main.rs:2:7: 292:4282133246 (#0), error=DistinctSources(DistinctSources { begin: (Real(LocalPath("src/main.rs")), BytePos(0)), end: (Real(Remapped { local_path: None, virtual_name: "/rustc/e90c5fbbc5df5c81267747daeb937d4e955ce6ad/library/unwind/src/libunwind.rs" }), BytePos(12823298)) })
error[E0435]: attempt to use a non-constant value in a constant
   --> src/main.rs:2:20
    |
2   |       |x: usize| [0; x];
    |         -            ^ non-constant value
    |  _______|
    | |
3   | |     // (note the space before "fn")
4   | | }
...   |

For more information about this error, try `rustc --explain E0435`.
error: could not compile `playground` due to previous error
```
</p>
</details>

This issue was found thanks to [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc), but the actual minimized code was hard to understand (` #![l=|x|[b;x`) so I unminimized it a bit. A related issue I found is that the span is wrong when there is whitespace between "let" and "x" here:

```rust
fn main() {
    let          x = 0;
    [0; x];
}
```

```
error[E0435]: attempt to use a non-constant value in a constant
 --> src/main.rs:3:9
  |
2 |     let          x = 0;
  |              ----- help: consider using `const` instead of `let`: `const x`
3 |     [0; x];
  |         ^ non-constant value
```

### PR DESCRIPTION
**Pull request by Noratrieb**
This also fixes the same suggestion, which was kind of broken, because it just searched for the last occurence of `const` to replace with a `let`. This works great in some cases, but when there is no const and a leading space to the file, it doesn't work and panic with overflow because it thought that it had found a const.

I also changed the suggestion to only trigger if the `const` and the non-constant value are on the same line, because if they aren't, the suggestion is very likely to be wrong.

Also don't trigger the suggestion if the found `const` is on line 0, because that triggers the ICE.

Asking Esteban to review since he was the last one to change the relevant code.

r? @estebank 

Fixes #90878

### CODE REVIEW

**Comment by Badel2** [comment]
Hi, I found another edge case that is not fixed by this change, when the first character is a newline instead of a space:

```

fn main() {
    |x: usize| [0; x];  //~ ERROR attempt to use a non-constant value in a constant [E0435]
    // (note the empty line before "fn")
}
```

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 1]
```diff
@@ -450,9 +450,23 @@ impl<'a> Resolver<'a> {
...
-                if sp.lo().0 == 0 {
+
+                let is_first_line = self
+                    .session
+                    .source_map()
+                    .lookup_line(sp.lo())
+                    .map(|file_and_line| file_and_line.line == 0)
+                    .unwrap_or(true);
+
+                if sp.lo().0 == 0 || self.session.source_map().is_multiline(sp) || is_first_line {
```
Comment (Badel2) [inline compiler/rustc_resolve/src/diagnostics.rs:469]: You can avoid the ICE by checking for overflow here, this removes the need to check the line number:

```suggestion
                if sp.lo().0.saturating_sub(current.len() as u32) == 0 || self.session.source_map().is_multiline(sp) {
```

**Comment by Nilstrieb** [comment]
I actually tested it out with newlines myself, but apparently I messed up, thank you for finding it!

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 1 above)
Comment (Nilstrieb) [reply compiler/rustc_resolve/src/diagnostics.rs:469]: That is certainly nicer and better, thank you!

**Comment by Nilstrieb** [comment]
I don't find a way to make a test for that case since tidy rejects the test because of the traling newline. But I tested it locally and let's hope that there will never be a regression for that case 😅 (if you know of an option to ignore tidy for a file like that hit me up)

**Comment by estebank** [comment]
`// ignore-tidy-end-whitespace` and/or `// ignore-tidy-trailing-newlines`

**Comment by Nilstrieb** [comment]
Does the ignore comment have to be at the start of the file or anywhere? Because it wouldn't be possible to insert it at the start. (also it's a leading newline and not a trailing one but I guess you can ignore that too)

**Comment by Nilstrieb** [comment]
Turns out there wasn't an option for ignoring leading newlines, so I simply added it :). I hope this is that misplaced here, but tell me how I should handle it otherwise if this isn't very good

**Pull request merged**